File: typescript/problem_76.ts
export function minColumnsToRemove(rows: string[]): number {
    if (!rows.length) return 0;
    let remove = 0;
    for (let c = 0; c < rows[0].length; c++)
        for (let r = 1; r < rows.length; r++)
            if (rows[r][c] < rows[r-1][c]) { remove++; break; }
    return remove;
}

export function problem76(): void {
    console.log("Problem 76:");
    if (minColumnsToRemove(["cba","daf","ghi"]) !== 1) throw new Error("assertion failed");
    if (minColumnsToRemove(["abcdef"])           !== 0) throw new Error("assertion failed");
    if (minColumnsToRemove(["zyx","wvu","tsr"])  !== 3) throw new Error("assertion failed");
    console.log("All Tests Passed");
}

File: rust/problem_76.rs
pub fn min_columns_to_remove(rows: &[&str]) -> usize {
    if rows.is_empty() { return 0; }
    let cols = rows[0].len();
    let row_bytes: Vec<&[u8]> = rows.iter().map(|s| s.as_bytes()).collect();
    (0..cols).filter(|&c| (1..rows.len()).any(|r| row_bytes[r][c] < row_bytes[r-1][c])).count()
}

pub fn problem76() {
    println!("Problem 76:");
    assert_eq!(min_columns_to_remove(&["cba","daf","ghi"]), 1);
    assert_eq!(min_columns_to_remove(&["abcdef"]),           0);
    assert_eq!(min_columns_to_remove(&["zyx","wvu","tsr"]), 3);
    println!("All Tests Passed");
}
